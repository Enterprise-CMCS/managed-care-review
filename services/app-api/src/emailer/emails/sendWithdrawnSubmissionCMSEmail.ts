import type { ContractType } from '../../domain-models'
import type { RateForDisplay } from '../../postgres/contractAndRates/withdrawContract'
import type {
    EmailConfiguration,
    EmailData,
    StateAnalystsEmails,
} from '../emailer'
import { pruneDuplicateEmails } from '../formatters'
import { rateSummaryURL, submissionSummaryURL } from '../generateURLs'
import { renderTemplate, stripHTMLFromTemplate } from '../templateHelpers'
import { findStatePrograms, packageName } from '@mc-review/hpp'
import { formatCalendarDate } from '@mc-review/dates'

type FormattedRateDisplayDataType = {
    rateCertificationName: string
    rateSummaryURL: string
}

export type withdrawContractEtaData = {
    contractName: string
    contractSummaryURL: string
    withdrawnBy: string
    withdrawnOn: string
    withdrawReason: string
    formattedRateDisplayData: FormattedRateDisplayDataType[]
}

export const parseAndValidateContractsAndRates = (
    contract: ContractType,
    rates: RateForDisplay[],
    config: EmailConfiguration
): withdrawContractEtaData | Error => {
    if (contract.consolidatedStatus !== 'WITHDRAWN') {
        return new Error('Contract consolidated status should be WITHDRAWN')
    }

    // Withdrawn Contract Name
    const contractName = packageName(
        contract.stateCode,
        contract.stateNumber,
        contract.packageSubmissions[0].contractRevision.formData.programIDs,
        findStatePrograms(contract.stateCode)
    )

    // Summary URL for the withdrawn submission
    const contractSummaryURL = submissionSummaryURL(contract.id, config.baseUrl)

    // Check to make sure withdrawn contract has actions
    const contractReviewStatusActions = contract.reviewStatusActions
    if (
        !contractReviewStatusActions ||
        contractReviewStatusActions.length === 0
    ) {
        return new Error('Contract does not have any review status actions')
    }

    // Latest stauts action gets us the withdrawn by info
    const latestStatusAction = contractReviewStatusActions.sort(
        (a, b) =>
            new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime()
    )[0]
    if (latestStatusAction.actionType !== 'WITHDRAW') {
        return new Error('Latest contract review action is not WITHDRAW')
    }

    // Withdraw info
    const withdrawnBy = latestStatusAction.updatedBy.email
    const withdrawnOn = formatCalendarDate(
        latestStatusAction.updatedAt,
        'America/Los_Angeles'
    )
    const withdrawReason = latestStatusAction.updatedReason!

    // Rates withdrawn with the submission that will be displayed in the email
    const formattedRateDisplayData: FormattedRateDisplayDataType[] = rates.map(
        (rate) => ({
            rateCertificationName: rate.rateCertificationName,
            rateSummaryURL: rateSummaryURL(rate.id, config.baseUrl),
        })
    )

    return {
        contractName,
        contractSummaryURL,
        withdrawnBy,
        withdrawnOn,
        withdrawReason,
        formattedRateDisplayData,
    }
}

export const sendWithdrawnSubmissionCMSEmail = async (
    withdrawnContract: ContractType,
    ratesForDisplay: RateForDisplay[],
    stateAnalystsEmails: StateAnalystsEmails,
    config: EmailConfiguration
): Promise<EmailData | Error> => {
    const toAddresses = pruneDuplicateEmails([
        ...stateAnalystsEmails,
        ...config.dmcpSubmissionEmails,
        ...config.oactEmails,
        ...config.devReviewTeamEmails,
    ])

    const etaData = parseAndValidateContractsAndRates(
        withdrawnContract,
        ratesForDisplay,
        config
    )
    if (etaData instanceof Error) {
        return etaData
    }

    const template = await renderTemplate(
        'sendWithdrawnSubmissionCMSEmail',
        etaData
    )

    if (template instanceof Error) {
        return template
    } else {
        return {
            toAddresses,
            replyToAddresses: [],
            sourceEmail: config.emailSource,
            subject: `${
                config.stage !== 'prod' ? `[${config.stage}] ` : ''
            }${etaData.contractName} was withdrawn by CMS`,
            bodyText: stripHTMLFromTemplate(template),
            bodyHTML: template,
        }
    }
}
