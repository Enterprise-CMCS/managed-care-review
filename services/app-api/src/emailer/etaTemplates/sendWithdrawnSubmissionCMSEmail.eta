<h1 style=<%~'"font-weight: normal;"'%>>Submission <%= it.contractName %> was withdrawn by CMS</h1>
<b>Withdrawn by:</b> <%= it.withdrawnBy %><br />
<b>Withdrawn on:</b> <%= it.withdrawnOn %><br />
<b>Reason for withdrawal:</b> <%= it.withdrawReason %><br />
<% if (it.formattedRateDisplayData && it.formattedRateDisplayData.length > 0) { %>
<b>Related rate certification(s):</b>
<ul>
<% it.formattedRateDisplayData.forEach(function(rate) { %>
  <li><%= rate.rateCertificationName %></li>
<% }) %>
</ul>
<% } %>
<% if (it.formattedRateDisplayData.length === 0) { %>
<br />
<% } %>
<a href="<%= it.contractSummaryURL %>">View submission in MC-Review</a>