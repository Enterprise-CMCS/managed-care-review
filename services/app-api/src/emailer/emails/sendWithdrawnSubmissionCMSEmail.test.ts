import {
    mockContract,
    testEmailConfig,
    testStateAnalystsEmails,
} from '../../testHelpers/emailerHelpers'
import { sendWithdrawnSubmissionCMSEmail } from './sendWithdrawnSubmissionCMSEmail'

const testContract = mockContract({
    id: 'test-contract-123',
    consolidatedStatus: 'WITHDRAWN',
    reviewStatusActions: [
        {
            updatedAt: new Date('2025-04-10'),
            updatedBy: {
                role: 'CMS_USER',
                email: 'zuko@example.com',
                givenName: 'Zuko',
                familyName: 'Hotman',
            },
            updatedReason: 'Test Withdraw',
            dateApprovalReleasedToState: undefined,
            actionType: 'WITHDRAW',
            contractID: 'test-contract-123',
        },
    ],
})

const testRatesForDisplay = [
    {
        id: 'rate-123',
        rateCertificationName: 'rate-name-123',
    },
    {
        id: 'rate-321',
        rateCertificationName: 'rate-name-321',
    },
]
const stateAnalystEmails = () => [...testStateAnalystsEmails]

describe('sendWithdrawnSubmissionCMSEmail', () => {
    it('CMS email contains correct toAddresses', async () => {
        const template = await sendWithdrawnSubmissionCMSEmail(
            testContract,
            testRatesForDisplay,
            stateAnalystEmails(),
            testEmailConfig()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template.toAddresses).toEqual(
            expect.arrayContaining([
                ...stateAnalystEmails(),
                ...testEmailConfig().dmcpSubmissionEmails,
                ...testEmailConfig().oactEmails,
            ])
        )
    })

    it('renders overall email for withdrawn contract as expected', async () => {
        const template = await sendWithdrawnSubmissionCMSEmail(
            testContract,
            testRatesForDisplay,
            stateAnalystEmails(),
            testEmailConfig()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template.bodyHTML).toMatchSnapshot()
    })
})

describe('sendWithdrawnSubmissionCMSEmail error handling', () => {
    it('returns an error if contract consolidated status is not WITHDRAWN', async () => {
        const contractWithWrongStatus = {
            ...testContract,
            consolidatedStatus: 'RESUBMITTED' as const,
        }

        const template = await sendWithdrawnSubmissionCMSEmail(
            contractWithWrongStatus,
            testRatesForDisplay,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Contract consolidated status should be WITHDRAWN'
        )
    })

    it('returns an error if review status actions is empty', async () => {
        const contractWithMissingReviewActions = {
            ...testContract,
            reviewStatusActions: [],
        }

        const template = await sendWithdrawnSubmissionCMSEmail(
            contractWithMissingReviewActions,
            testRatesForDisplay,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Latest contract review action is not WITHDRAW'
        )
    })

    it('returns an error if the latest contract review action is not WITHDRAW', async () => {
        const contractWithWrongActionType = {
            ...testContract,
            reviewStatusActions: [
                {
                    updatedAt: new Date('2025-04-10'),
                    updatedBy: {
                        role: 'CMS_USER' as const,
                        email: 'zuko@example.com',
                        givenName: 'Zuko',
                        familyName: 'Hotman',
                    },
                    updatedReason: 'Test Withdraw',
                    dateApprovalReleasedToState: undefined,
                    actionType: 'UNDER_REVIEW' as const,
                    contractID: 'test-contract-123',
                },
            ],
        }

        const template = await sendWithdrawnSubmissionCMSEmail(
            contractWithWrongActionType,
            testRatesForDisplay,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Latest contract review action is not WITHDRAW'
        )
    })
})
