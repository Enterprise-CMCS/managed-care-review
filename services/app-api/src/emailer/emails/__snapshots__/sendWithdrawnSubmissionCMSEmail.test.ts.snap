// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sendWithdrawnSubmissionCMSEmail > renders overall email for withdrawn contract as expected 1`] = `
"<h1 style="font-weight: normal;">Submission MCR-MN-0004-SNBC was withdrawn by CMS</h1>
<b>Withdrawn by:</b> zuko@example.com<br />
<b>Withdrawn on:</b> 04/09/2025<br />
<b>Reason for withdrawal:</b> Test Withdraw<br />
<b>Related rate certification(s):</b>
<ul>
  <li><a href="http://localhost/rates/rate-123">rate-name-123</a></li>
  <li><a href="http://localhost/rates/rate-321">rate-name-321</a></li>
</ul>
<a href="http://localhost/submissions/test-contract-123">View submission in MC-Review</a>
"
`;
